search-provider/yelp-search-provider.c

```diff
@@ -0,0 +1,498 @@
+                                         "org.gnome.Yelp.SearchProvider",
+                                         G_BUS_NAME_OWNER_FLAGS_NONE,
+                                         search_provider_bus_acquired_cb,
+                                         search_provider_name_acquired_cb,
+                                         search_provider_name_lost_cb,
+                                         app, NULL);
+    g_free (page_ids);
+    app->doc = g_object_ref (document);
+}
+
+static void
+uri_resolved_cb (YelpUri  *uri,
+                 YelpSearchProviderApp *app)
+{
+    YelpDocument *doc;
+    gchar *page_id;
+
+    doc = yelp_document_get_for_uri (uri);
+    g_signal_handler_disconnect (uri, app->id);
+    page_id = yelp_uri_get_page_id (uri);
+    yelp_document_request_page (doc,
+                               page_id,
+                               NULL,
+                               (YelpDocumentCallback) preload_data_cb,
+                               app,
+                               NULL);
+
+    g_object_unref (uri);
+    g_object_unref (doc);
+}
+
+static void
+search_provider_app_startup (GApplication *app)
+{
+    YelpSearchProviderApp *self = YELP_SEARCH_PROVIDER_APP (app);
+    YelpUri *base_uri;
+    G_APPLICATION_CLASS (yelp_search_provider_app_parent_class)->startup (app);
+
+    // TODO: remove holding the app forever
+    if (g_getenv ("YELP_SEARCH_PROVIDER_PERSIST") != NULL)
+        g_application_hold (app);
+
+    base_uri = yelp_uri_new ("help:gnome-help");
+    self->id = g_signal_connect (base_uri,
+                                 "resolved",
+                                 G_CALLBACK (uri_resolved_cb),
+                                 self);
+
+    self->page_data_hash_map = g_hash_table_new_full (g_str_hash,
+                                                      g_str_equal,
+                                                      g_free,
+                                                      (GDestroyNotify) page_data_free);
+    yelp_uri_resolve (base_uri);
+}
+
+static void
+yelp_search_provider_app_init (YelpSearchProviderApp *self)
+{
+    g_application_set_inactivity_timeout (self, SEARCH_PROVIDER_INACTIVITY_TIMEOUT);
+    g_application_hold(self);
+}
+
+static void
+yelp_search_provider_app_class_init (YelpSearchProviderAppClass *klass)
+{
+    GApplicationClass *aclass = G_APPLICATION_CLASS (klass);
+    GObjectClass *oclass = G_OBJECT_CLASS (klass);
+
+    aclass->startup = search_provider_app_startup;
+    oclass->dispose = search_provider_app_dispose;
+}
+
+static GApplication *
+yelp_search_provider_app_new (void)
+{
+    return g_object_new (YELP_TYPE_SEARCH_PROVIDER_APP,
+                         "application-id", "org.gnome.Yelp.SearchProvider",
+                         "flags", G_APPLICATION_IS_SERVICE,
+                         NULL); 
+}
+
+int
+main (int argc,
+      char **argv)
+{
+    GApplication *app;
+    gint res;
+    gchar *env_var = "YELP_SEARCH_PROVIDER";
```

Comment:
Is this still needed? I think that calling `gtk_init` below fixed the issue with settings stuff, no?